OC-1532 Added support for backing up swift containers.

jbzdak opened the pull request:
# Testing instructions

## Installation 

I will prepare Ansible scripts that deploy backup functionalities on a production environment, however for manual tests it might be easier to use devstack/IM vagrant VMV. Setting up devstack is somewhat cumbersome, if this turns out to be taking a lot of your time, please wait until I prepare a IM sandbox for this feature. 

1. [Optional step] Take a snapshot, or install atop a fresh VM. 
2. Pull this branch.
3. Download roles ``cd deploy && ansible-galaxy install -r requirements.yml``. 
3. Install tarsnap on the devstack. To install tarsnap you'll need to run ansible scripts on the VM. First get connection details ``vagrant ssh-config``, cd to deploy folder and `ansible-playbook install_tarsnap.yml -i localhost, -e "ansible_port=2222" --ssh-common-args "-o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no" -vvvv -u vagrant -k`. This will ask for password which is `vagrant`. Minor details (like port) might vary. 
5. Add following to `.env`: 

```
BACKUP_SWIFT_ENABLED = true

BACKUP_SWIFT_TARGET = /var/cache/swift-data-backup
BACKUP_SWIFT_TARSNAP_CACHE_LOCATION = /var/cache/tarsnap
BACKUP_SWIFT_TARSNAP_KEY_LOCATION = /home/vagrant/opencraft/tarsnap.key
```
5. Create /var/cache/tarsnap and /var/cache/swift-data-backup on the VM and make vagrant user an owner. 
4. Download tarsnap key attached to OC-1532, and copy it as `tarsnap.key` in the IM folder. 

## Testing. 

1. Log in to the VM and run `make manage backup_swift`. Apart from testing backups. This will test whether: we can recover errors with tarsnap cache (which is entirely missing). You should get "Got tarsnap error. Trying to fix it by running tarsnap --fsck". 
2. Run `make manage backup_swift`, `make manage backup_swift` once again. Now  "Got tarsnap error. Trying to fix it by running tarsnap --fsck", but you'll get: `Object GET failed: https://storage.bhs1.cloud.ovh.net/v1/AUTH_d63f739f98604cb799e584eebbb6057d/safe-to-delete-once-OC-1532-is-merged/canary?multipart-manifest=get 304 Not Modified`. This will be explained shortly. 
3. Test that backup task will be run from cron. Either set time on the VM to `1:09 AM` or go to `backup_swift/tasks.py` and change cron to run this every minute. 
4. Download tarsnap backup and verify the contents. To verify tarsnap backup you'll need to: 

  1. List tarsnap archives: `tarsnap --keyfile /home/vagrant/opencraft/tarsnap.key --cachedir /var/cache/tarsnap/ --list-archives | sort`, this will produce a list of time-stamped archives. 
  2. List archive contents: `tarsnap --keyfile /home/vagrant/opencraft/tarsnap.key --cachedir /var/cache/tarsnap/ -t -f im-swift-backup-2016-06-03T05:46:47.557570` 
  3. Download the archive `cd /tmp && tarsnap --keyfile /home/vagrant/opencraft/tarsnap.key --cachedir /var/cache/tarsnap/ -x -f im-swift-backup-2016-06-03T05:46:47.557570`  (the only difference between this and the previous command is `-x` instead of `-t`)

5. To test snitch I think that it would be easiest to: create new snitch at https://deadmanssnitch.com/ (credentials are in the usual place). Set the url as `BACKUP_SWIFT_SNITCH` and run command manually --- snitch should register the backup. After that just delete the snitch from the https://deadmanssnitch.com/ 

# Pending issues/TODO/ Authors concerns

- [x] Need to rebase
- [x] Need to squash commits
- [x] Due to some arguable design decisions, `python-swiftclient` decides to handle `304` HTTP statuses when downloading files as errors, that need to be handled (they are) and that are logged. Since I wanted to conserve bandwidth, and make downloading of files faster I decided to use the `skip_identical` option for swift --- which computes `MD5` for the files, and skips them if MD5 match, or rather --- if they match we get 304, this would produce quite a lot of log errors, that need to be suppressed (or this needs to be solved another way)
- [x] Need to add ansible scripts that will set up production server with tarsnap.

jbzdak commented on backup_swift/management/commands/backup_swift.py:
```diff
@@ -0,0 +1,34 @@
+    """
+    Management command that runs backup swift task.
+
+    Mostly used as to speed up manual tests of the code.
```
This is mostly here to enable me to test backups fast.

jbzdak commented on instance/openstack.py:
```diff
@@ -99,39 +103,88 @@ def get_server_public_address(server):
 
 
-def get_swift_connection(
+def swift_service(
```
`python-switclient` contains two separate API's to interact with swift. One is request oriented (one call is one request) and other one is focused on batch operations. I needed to use the "batch" operation's one, so it made sense to rewrite existing calls to this new service.

pomegranited commented on .env.test:
```diff
@@ -15,4 +15,4 @@ GANDI_API_KEY='TEST_GANDI_API_KEY'
 WATCH_ORGANIZATION='test-org'
 WATCH_FORK='watched/fork'
-BASE_HANDLERS='["file"]'
+BASE_HANDLERS='["file"]'
```
nit: unnecessary change

pomegranited commented on backup_swift/tarsnap.py:
```diff
@@ -0,0 +1,100 @@
+
+    logger.info("Got tarsnap error. Trying to fix it by running tarsnap --fsck")
+    run_tarsnap_command(fsck_command)
+    run_tarsnap_command(create_archive_command)
```
nit: No need for `return`/`pass`?
```python
    try:
        run_tarsnap_command(create_archive_command)
    except TarsnapFsckException:
        logger.info("Got tarsnap error. Trying to fix it by running tarsnap --fsck")
        run_tarsnap_command(fsck_command)
        run_tarsnap_command(create_archive_command)
```

jbzdak replied:
Thanks! Fixed in: ad826900e4e2a64c819bd22da51515484466aec7

pomegranited commented on backup_swift/tasks.py:
```diff
@@ -0,0 +1,120 @@
+
+    # In case of downloading errors run tarsnap nevertheless, so we at least backup something.
+
+    tarsnap_succesfull = make_tarsnap_backup(
```
nit: it's spelled "successful"

jbzdak replied:
Fixed in:

pomegranited commented on backup_swift/tasks.py:
```diff
@@ -0,0 +1,120 @@
+    )
+
+    if not tarsnap_succesfull:
+        error_report += "Error while running tarsnap\n"
```
It's worth pointing the admins who may see this email to the server log files for detailed error messages.

jbzdak replied:
Fixed in: ad826900e4e2a64c819bd22da51515484466aec7

jbzdak replied:
Reverted in: ed978f455e251581b5e9458952ab9da309ea0720

pomegranited commented on backup_swift/tasks.py:
```diff
@@ -0,0 +1,120 @@
+            ping_heartbeat_url(settings.BACKUP_SWIFT_SNITCH)
+
+
+@db_task()
```
Since this task doesn't execute any DB queries (as far as I can see?), can it just be an `@task()` ?

jbzdak replied:
Thanks! At some point I wanted it to access instance model, but then it wasn't really needed. Fixed in: ad826900e4e2a64c819bd22da51515484466aec7

jbzdak replied:
Reverted in: ed978f455e251581b5e9458952ab9da309ea0720

pomegranited commented on backup_swift/tasks.py:
```diff
@@ -0,0 +1,120 @@
+
+if settings.BACKUP_SWIFT_ENABLED:
+
+    @db_periodic_task(crontab(minute="10", hour="1"))
```
Since this task doesn't execute any DB queries, can it just be an `@periodic_task(..)` ?

jbzdak replied:
Fixed in: ad826900e4e2a64c819bd22da51515484466aec7

jbzdak replied:
Ech. Looks like every single task in IM uses database --- to log to database. I'll change the decorators back.

jbzdak replied:
Reverted in: ed978f455e251581b5e9458952ab9da309ea0720

pomegranited replied:
Sorry @jbzdak !

pomegranited commented on backup_swift/tasks.py:
```diff
@@ -0,0 +1,120 @@
+
+        This is long running task, more like spawn_appserver than watch_pr,
+        since we a single queue for periodic tasks it makes sense for this to finish
+        early and then execute in another worker.
```
Makes sense, thanks for the explanation :)

pomegranited commented:
@jbzdak Your changes are looking good so far 😄   I don't know much about tarsnap or swift storage, but have been able to follow your logic very clearly.

Couple of questions:

* Step 3 above is missing a step, before you can run the `install_tarsnap.yml`: `ansible-galaxy install pmbauer.tarsnap`
* I was able to run `make manage backup_swift` manually in my devstack, but was unable to see the cron task run.  I've done:
    * enabled the settings you indicated
    * updated the cron time in `backup_swift/tasks.py` to run every minute
    * ran `make run`.
    There I saw `backup_swift_periodic` in the list of available commands, and saw huey workers run `pr_watch`, but `backup_swift_periodic` never got executed.
* Where can I find the tarsnap file?  I didn't find anything named `im-swift-backup` on my devstack.
* How do I test `BACKUP_SWIFT_SNITCH` ?

jbzdak commented:
@pomegranited I have updated instructions so commands required to verify tarsnap backup are included.

jbzdak commented:
@pomegranited  I have rebased this on top of master. And addressed all your concerns, as to running this as a periodic task, this task is scheduled daily, however when I set it to execute every minute by: `@db_periodic_task(crontab(minute="*", hour="*"))` it scheduled properly. 

I have also updated testing instructions to describe how to test tarsnap and snitch.

antoviaque commented on deploy/install_tarsnap.yml:
```diff
@@ -0,0 +1,17 @@
+# Intended to install tarsnap on the remote machine, can be removed once we rewrite
+# bin/bootstrap to use ansible.
```
Can you add this to the scope of OC-1591 to make sure we don't forget?

jbzdak replied:
Thanks! Updated OC-1591.

pomegranited commented:
@jbzdak  Your updated code, tarsnap, and Dead Man's Snitch details check out.  Thank you for the detailed instructions!

👍  once tests pass.
